sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala

```diff
@@ -215,7 +216,7 @@ private[parquet] class RowWriteSupport extends WriteSupport[InternalRow] with Lo
           if (d.precisionInfo == None || d.precisionInfo.get.precision > 18) {
             sys.error(s"Unsupported datatype $d, cannot write to consumer")
           }
-          writeDecimal(value.asInstanceOf[Decimal], d.precisionInfo.get.precision)
+          writeDecimal(convertDecimal(value), d.precisionInfo.get.precision)
```

Comment:
Then can you write your tests in python instead of scala? As I can't reproduce it with scala...